ds701_book/27-RNN.qmd

```diff
@@ -0,0 +1,640 @@
+
+## Addressing RNN Limitations
+
+Some proposed solutions for mitigating the previous issues are
+
+- **LSTM (Long Short-Term Memory)**
+- **GRU (Gated Recurrent Units)** 
+
+These are more advanced variants of RNNs designed to address some of these limitations by improving the ability to capture long-term dependencies and mitigating gradient issues.
+
+## LSTM
+
+Below is an illustration of the LSTM architecture.
+
+![](drawio/LSTM.png)
+
+---
+
+The new components are:
+
+- The input cell state $c_{t-1}$ and output cell state $c_{t}$.
+- Yellow blocks consisting of neural network layers with sigmoid $\sigma$ or tanh activation functions.
+- Red circles indicating point-wise operations.
+
+
+## Cell State
+
+:::: {.columns}
+::: {.column width="50%"}
+
+The cell state $c_{t-1}$ is the input to the LSTM block.
+
+This value then moves through the LSTM block.
+
+It is modified by either a multiplication or addition interaction. 
+
+After these operations, the modified input state becomes $c_{t}$ and is sent to the next LSTM block.
+
+:::
+::: {.column width="50%"}
+![](drawio/CellState.png)
+:::
+::::
+
+## Forget Gate Layer
+
+:::: {.columns}
+::: {.column width="50%"}
+The forget layer computes a value 
+
+$$
+f_t = \sigma(W_f[h_{t-1}, x_t] + b_f),
+$$
+
+where $\sigma$ is the sigmoid function.
+
+The output $f_t$ is a vector of numbers between 0 and 1. These values multiply the corresponding vector values in $C_{t-1}$.
+
+A value of 0 says throw that component of $C_{t-1}$ away. A value of 1 says keep that component of $C_{t-1}$.
+
+This operation tells us which *old* information in the cell state we should keep or remove.
+
+:::
+::: {.column width="50%"}
+![](drawio/ForgetLayer.png)
+:::
+::::
+
+## Input Gate Layer
+
+:::: {.columns}
+::: {.column width="50%"}
+The input gate layer computes values
+
+$$
+\begin{align}
+i_t &= \sigma(W_i[h_{t-1}, x_t] + b_i), \\
+\tilde{c}_{t} &= \operatorname{tanh}(W_c[h_t-1, x_t] + b_c).
+\end{align}
+$$
+
+The value that is added to the cell state $c_{t-1}$ is $i_t \cdot \tilde{c}_t$.
+
+This value tells us what *new* information to add to the cell state.
+
+At this stage of the process, the cell state now has the formula
+
+$$
+c_t = f_t \cdot c_{t-1} + i_t \cdot \tilde{c}_t.
+$$
+
+:::
+::: {.column width="50%"}
+![](drawio/InputGateLayer.png)
+:::
+::::
+
+## Output Layer
+
+:::: {.columns}
+::: {.column width="50%"}
+The output gate layer computes values
+
+$$
+\begin{align}
+o_t &= \sigma(W_o[h_{t-1}, x_t] + b_o), \\
+\tilde{c}_{t} &=o_t * \operatorname{tanh}(c_t).
+\end{align}
+$$
+
+The vector $o_t$ from the sigmoid layer tells us what parts of the cell state we use for output. The output is a filtered version of the hyperbolic tangent of cell state $c_t$. 
+
+The output of the block is $y_t$. It is the same as the hidden state $h_t$ that is sent to the $t+1$ LSTM block.
+
+This value tells us what *new* information to add to the cell state.
+
```

Comment:
Perhaps list the total trainable weight matrices and bias vectors. 4 of each, right?

Reply:
Added a weight summary slide.